Cookbook/Proposals/MultilineArgumentsBracket_SwiftLint_rule.md

```diff
@@ -0,0 +1,121 @@
+# Enable multiline rules in SwiftLint
+
+* Author(s): Adam Borek
+* Review Manager: Diego Petrucci
+
+## Introduction
+The motivation to create this proposal to not waste the time to fix violations of a SwfitLint's rule which we maybe wouldn't like to have, considering all pros & cons.
+
+## Motivation
+In our [StyleGuide](https://github.com/Babylonpartners/ios-playbook/tree/master/Cookbook/Style-guide#function-declarations) we have a section about multiline arguments & parameters. Multiline function declaration or function calls should have their brackets in separate lines e.g:
+```swift
+func reticulateSplines(
+  spline: [Double],
+  adjustmentFactor: Double,
+  translateConstant: Int,
+  comment: String
+) -> Bool {
+  // reticulate code goes here
+}
+
+let success = reticulateSplines(
+  spline: splines,
+  adjustmentFactor: 1.3,
+  translateConstant: 2,
+  comment: "normalize the display"
+)
+```
+To reduce our time spent on fixing such nits we can use SwiftLint to warns us about such issues.
+
+## Proposed solution
+Turn on following SwiftLint rules:
+```
+- multiline_arguments
+- multiline_arguments_brackets
+- multiline_parameters
+- multiline_parameters_brackets
+```
+The documentation with samples what particular rule does can be found [here](https://github.com/realm/SwiftLint/blob/master/Rules.md#multiline-arguments).
+
+## Impact on existing codebase
+After adding rules to our codebase, Xcode reports about 2500+ violations of them. That would need to be solved before marging to develop. More than the half of the work can be automated by using SwiftFormat to automatically format files. After running [SwiftFormat](https://github.com/nicklockwood/SwiftFormat) there's around ~900 issues to be fixed manually.
+
+In my opinion the impact on the codebase will be very positive, however it will also add few changes which are not defined in our style guide. I think it's acceptable price to make it possible to resolve nits quicker which would decrease a time when a PR is open.
+
+However, the given set of rules goes a step further. `multiline_arguments_brackets` rule makes it possible achieve what we want, which is, having a trailing bracket at a function call in the new line `)`. However, this rule **always** requires to put trailing bracket `)` in a newline if the function takes more than just one line to be invoked. Even if it's only one argument.
```

Comment:
> Even if it's only one argument.

Can you add an example of what it means on practice?

Reply:
@ilyapuchka there are examples beloew ;)

Reply:
In this case it might collide with code style for arrays, so I was looking for example that would be free of other rules. Will this:

```swift
func foo(bar: Bar)
```

always trigger violation and will require to change it to this?

```swift
func foo(
  bar: Bar
)
```

I'm not necessary against this edge case.

Reply:
@ilyapuchka When I read the proposal again I think I know what has mislead you to ask that question ;)
Here is updated version of one sentence :

> Few examples what pieces of the code trigger the rule and how can this be fixed. Keep in mind this is a list of changes which are not defined in the Function Definition or Function Call sections of the StyleGuide.

TL;DR All examples I've shown in the proposal are changes which don't come from the `Function Definition` or `Function Call` rules from our Style Guide. Those are examples of changes which are also forced on us to bring to our Style Guide if we decide to merge the proposal.

Reply:
So what you are trying to say with this note then is that 

```swift
foo(bar: Bar())
```

or 

```swift
foo(
  bar: Bar()
)
``` 

will be fine, but 

```swift
foo(
   bar: Bar())
```

will be not, right?

Reply:
I've updated the proposal. I hope now it's more on point ;)

My problem is the following. What we want to achieve is to have function parameters/arguments defined in separate lines as well to have brackets in separate lines:

```swift
func reticulateSplines(
  spline: [Double],
  adjustmentFactor: Double,
  translateConstant: Int,
  comment: String
) -> Bool {
  // reticulate code goes here
}

let success = reticulateSplines(
  spline: splines,
  adjustmentFactor: 1.3,
  translateConstant: 2,
  comment: "normalize the display"
)
```

SwiftLint has a rule for that but the rule also require changes in places which are not defined in our style guide. Examples `Trailing closure argument` `Inlined array argument` `Multiline inner argument ` show when **additional** changes would be required.